adam/perception/perception_graph.py

```diff
@@ -2932,10 +2946,168 @@ def _translate_frames(
+                    )
+            else:
+                # This edge does not also appear in the first frame,
+                # so no merging is needed.
+                _dynamic_digraph.add_edge(source, target, label=after_label)
+
+        decoded_object_to_cluster_node = {
+            object_cluster_node.cluster_id: object_cluster_node
+            for object_cluster_node in _dynamic_digraph.nodes
+            if isinstance(object_cluster_node, ObjectClusterNode)
+        }
+
+        with open(
+            perceptual_representation.simulated_actions_file, encoding="utf-8"
+        ) as action_feature_file:
+            action_features = yaml.safe_load(action_feature_file)
+            action_stroke_graph = action_features["stroke_graph"]
+
+            # This needs to look up the nodes in the original graphs?
+            agent_node = decoded_object_to_cluster_node[action_features["object_name"]]
+            involved_object_nodes = [
+                decoded_object_to_cluster_node[object_name]
+                for object_name in action_features["objects"]
+            ]
+            involved_object_nodes.append(agent_node)
+
+            action_decode_node = CategoricalNode(
+                label="action_gnn",
+                value=action_stroke_graph["action_name"].lower(),
+                weight=action_stroke_graph["confidence_score"],
+            )
+
+            joint_point_nodes = []
+            for i, joint_feature in enumerate(
+                zip(
+                    action_stroke_graph["joint_points_world_coords"],
+                    action_stroke_graph["joint_points_xyd_coords"],
+                    action_stroke_graph["joint_points_confidence_score"],
+                )
+            ):
+                world_coord, xyd_coord, conf = joint_feature
+                joint_point_nodes.append(
+                    JointPointNode(
+                        scene_xyd_coord=DepthPoint(
+                            xyd_coord[0], xyd_coord[1], xyd_coord[2]
+                        ),
+                        world_coord=Point(world_coord[0], world_coord[1], world_coord[2]),
+                        confidence=conf,
+                        weight=conf,
+                        temporal_index=i,
+                    )
+                )
+
+            # First we link the objects to the action node
+            edges_to_add: List[Tuple[GraphNode, GraphNode, OntologyNode]] = [
+                (object_node, action_decode_node, INVOLVED_IN_ACTION)
+                for object_node in involved_object_nodes
+            ]
+            edges_to_add.extend(
+                (action_decode_node, joint_node, JOINT_POINT_TRAVERSAL)
+                for joint_node in joint_point_nodes
+            )
+            edges_to_add.extend(
+                (agent_node, joint_node, HAS_PROPERTY_LABEL)
+                for joint_node in joint_point_nodes[1:-1]
+            )
```

Comment:
This section of code reads weird to me, I think because it's grouping several mostly-unrelated things. There are lots of edges we add that don't belong in this list, and these aren't obviously related to each other -- why these? The answer is "they have the same scope" but that requires reading the later code and isn't very satisfying.

I think it would be clearer to define three separate lists, say `object_involvement_edges`, `joint_point_traversal_edges`, and `agent_has_joints_edges`, and chain the lists together at the point where we create the `DURING` edges.

This sort of ties in with breaking this section up into smaller helpers (see comment 🍌).